src/mod_webdav.c

```diff
@@ -1207,11 +1208,18 @@ static int webdav_parse_chunkqueue(server *srv, connection *con, handler_ctx *hc
 #endif
 
 #ifdef USE_LOCKS
-static int webdav_lockdiscovery(connection *con, buffer *locktoken, const char *lockscope, const char *locktype, int depth) {
+static int webdav_lockdiscovery(connection *con, buffer *locktoken, const char *lockscope, const char *locktype, const char *owner, int depth) {
 
 	buffer *b = chunkqueue_append_buffer_open(con->write_queue);
 
-	http_header_response_set(con, HTTP_HEADER_OTHER, CONST_STR_LEN("Lock-Token"), CONST_BUF_LEN(locktoken));
+	/* wrap < > around the Lock-Token header */
+	size_t locktokenlen = buffer_string_length(locktoken);
+	char locktokenheader[locktokenlen + 3];
+	locktokenheader[0] = '<';
+	strcpy(locktokenheader + 1, locktoken->ptr);
```

Comment:
Better, but please never, never use strcpy().  Use memcpy() with locktokenlen to show that you know string length.  Also, the terminating '\0' is not required since the string length is passed to http_header_response_set().

Reply:
I think it's better to keep the terminating '\0' to avoid any possible buffer overread due to misuse.

Reply:
The buffer is on the stack and is single use, calling a single API that was written to take a length so that substrings could be passed without terminating '\0'.  If you think you know better with your generic platitude, then you have not looked carefully at the code, which is well scoped, or at the API you are calling and into which you are passing that buffer.  On the other hand, you must have looked at the API to know whether or not you needed to pass in owned (allocated) memory or not, so your generic statement does not apply here.

If you wonder why I am so particular, please perform this test: benchmark your original snprintf() code 1 million times and compare it to copying into buffer with a few extra lines of code, also run 1 million times.  (In both cases, make sure that the optimizer does not optimize the loop away.)